nixos/modules/services/logging/logrotate.nix

```diff
@@ -82,27 +83,67 @@ let
         '';
       };
     };
-
-    config.name = name;
   };
 
+  generateLine = k: v:
+    # special keys handled separately for backwards compatibility
+    if builtins.elem k ["name" "path" "priority" "enable" "user" "group" "keep"] then ""
+    else if k == "extraConfig" then "${v}"
+    else if k == "frequency" then "${v}\n"
+    else if k == "keep" then "rotate ${toString v}\n"
+    else if builtins.elem k ["firstaction" "lastaction" "prerotate" "postrotate" "preremove"]
+         then "${k}\n  ${v}\nendscript\n"
+    else if v == null then ""
+    else if isInt v then "${k} ${toString v}\n"
+    else if v == true then "${k}\n"
+    else if v == false then "no${k}\n"
+    else "${k} ${v}\n";
+  generateSection = settings: concatStringsSep "" (
+    (mapAttrsToList generateLine settings)
+    ++ [
+        # user and group must be handled together
+        (if (settings.user or null != null && settings.group or null != null)
+          then "su ${settings.user} ${settings.group}\n"
+          else "")
+        # if rotate has not been set use compat value from keep
+        (if (settings.rotate or null == null && settings ? keep)
+          then generateLine "rotate" settings.keep
+          else "")
+      ]
+  );
+
+  # generateSection includes a final newline hence weird closing brace
   mkConf = pathOpts: ''
-    # generated by NixOS using the `services.logrotate.paths.${pathOpts.name}` attribute set
     ${concatMapStringsSep " " (path: ''"${path}"'') (toList pathOpts.path)} {
-      ${optionalString (pathOpts.user != null || pathOpts.group != null) "su ${pathOpts.user} ${pathOpts.group}"}
-      ${pathOpts.frequency}
-      rotate ${toString pathOpts.keep}
-      ${pathOpts.extraConfig}
-    }
+    ${generateSection pathOpts}}
   '';
 
   paths = sortProperties (attrValues (filterAttrs (_: pathOpts: pathOpts.enable) cfg.paths));
-  configFile = pkgs.writeText "logrotate.conf" (
-    concatStringsSep "\n" (
-      [ "missingok" "notifempty" cfg.extraConfig ] ++ (map mkConf paths)
-    )
-  );
+  configFile = pkgs.writeTextFile {
+    name = "logrotate.conf";
+    text = concatStringsSep "\n" (
+      [ (generateSection (
+          { "missingok" = true; "notifempty" = true; }
+          // cfg.settings))
+      ] ++ (map mkConf paths)
+    );
+    checkPhase = optionalString cfg.checkConfig ''
+      # logrotate --debug also checks that users specified in config
+      # file exist, but we only have sandboxed users here so brown these
+      # out. according to man page that means su, create and createolddir.
+      # XXX find a way to use real /etc/passwd and /etc/groups
```

Comment:
I don't think it's possible to use the real `/etc/passwd` and `/etc/groups` at build time, especially when `mutableUsers == true`.